odo/backends/avro.py

```diff
@@ -0,0 +1,252 @@
+                #Otherwise, reraise exception
+                if exc.errno != errno.ENOENT:
+                    raise exc
+                return None
+
+    @staticmethod
+    def _get_append_writer(uri, writers_schema=None):
+        """
+        Returns an isntance of avro.datafile.DataFileWriter for appending
+        to an existing avro file at `uri`.  Does not take a writers schema,
+        because avro library requires that writers_schema embedded in existing
+        file be used for appending.
+
+        Parameters
+        ----------
+
+        uri : str
+            uri of avro existing, non-empty avro file
+
+        writers_schema : avro.schema.Schema object
+            If not None, checks that writers_schema in existing file is the same as supplied schema.
+            Avro does not allow writing records to a container file with multiple writers_schema.
+
+        Returns
+        -------
+        avro.datafile.DataFileWriter
+        """
+        rec_writer = io.DatumWriter()
+        df_writer = datafile.DataFileWriter(
+            open(uri, 'ab+'),
+            rec_writer
+        )
+        #Check for embedded schema to ensure existing file is an avro file.
+        embedded_schema = schema.parse(df_writer.get_meta('avro.schema'))
+
+        #If writers_schema supplied, check for equality with embedded schema.
+        if writers_schema:
+            assert embedded_schema == writers_schema, \
+                "writers_schema embedded in {uri} differs from user supplied schema for appending."
+
+        return df_writer
+
+    @staticmethod
+    def _get_new_writer(uri, sch):
+        """
+        Returns an isntance of avro.datafile.DataFileWriter for writing
+        to a new avro file at `uri`.
+
+        Parameters
+        ----------
+
+        uri : str
+            uri of avro existing, non-empty avro file
+
+        sch : avro.schema.Schema object
+
+        Returns
+        -------
+        avro.datafile.DataFileWriter
+        """
+        rec_writer = io.DatumWriter()
+        df_writer = datafile.DataFileWriter(
+            open(uri, 'wb'),
+            rec_writer,
+            writers_schema = sch
+        )
+        return df_writer
+
+    @property
+    def writer(self):
+        if hasattr(self, '_writer'):
+            return self._writer
+        else:
+            if os.path.exists(self.uri) and os.path.getsize(self.uri) > 0:
+                df_writer = self._get_append_writer(self.uri, self.schema)
+            else:
+                df_writer = self._get_new_writer(self.uri, self.schema)
+            self._writer = df_writer
+        return df_writer
+
+    def flush(self):
+        if hasattr(self, '_writer'):
+            self._writer.close()
+            del(self._writer)
+
+
+@resource.register('.+\.(avro)')
+def resource_avro(uri, schema=None, **kwargs):
+    return AVRO(uri, schema=schema, **kwargs)
+
+def discover_schema(sch):
+    if isinstance(sch, schema.RecordSchema):
```

Comment:
You could potentially use multiple dispatch here to avoid the large number of `isinstance` checks.